homeassistant/components/dlna_dmr/media_player.py

```diff
@@ -24,298 +27,499 @@
+            )
+        )
 
-    # wrap with DmrDevice
-    dlna_device = DmrDevice(upnp_device, event_handler)
+    async def async_will_remove_from_hass(self) -> None:
+        """Handle removal."""
+        for callback in self._remove_ssdp_callbacks:
+            callback()
+        self._remove_ssdp_callbacks.clear()
+        await self._device_disconnect()
+
+    async def async_ssdp_callback(
+        self, info: Mapping[str, Any], change: ssdp.SsdpChange
+    ) -> None:
+        """Handle notification from SSDP of device state change."""
+        _LOGGER.debug(
+            "SSDP %s notification of device %s at %s",
+            change,
+            info[ssdp.ATTR_SSDP_USN],
+            info.get(ssdp.ATTR_SSDP_LOCATION),
+        )
 
-    # create our own device
-    device = DlnaDmrDevice(dlna_device, name)
-    _LOGGER.debug("Adding device: %s", device)
-    async_add_entities([device], True)
+        try:
+            bootid_str = info[ssdp.ATTR_SSDP_BOOTID]
+            bootid: int | None = int(bootid_str, 10)
+        except (KeyError, ValueError):
+            bootid = None
+
+        if change == ssdp.SsdpChange.UPDATE:
+            # This is an announcement that bootid is about to change
+            if self._bootid is not None and self._bootid == bootid:
+                # Store the new value (because our old value matches) so that we
+                # can ignore subsequent ssdp:alive messages
+                try:
+                    next_bootid_str = info[ssdp.ATTR_SSDP_NEXTBOOTID]
+                    self._bootid = int(next_bootid_str, 10)
+                except (KeyError, ValueError):
+                    pass
+            # Nothing left to do until ssdp:alive comes through
+            return
 
+        if self._bootid is not None and self._bootid != bootid and self._device:
+            # Device has rebooted, drop existing connection and maybe reconnect
+            await self._device_disconnect()
+        self._bootid = bootid
 
-class DlnaDmrDevice(MediaPlayerEntity):
-    """Representation of a DLNA DMR device."""
+        if change == ssdp.SsdpChange.BYEBYE and self._device:
+            # Device is going away, disconnect
+            await self._device_disconnect()
 
-    def __init__(self, dmr_device, name=None):
-        """Initialize DLNA DMR device."""
-        self._device = dmr_device
-        self._name = name
+        if change == ssdp.SsdpChange.ALIVE and not self._device:
+            location = info[ssdp.ATTR_SSDP_LOCATION]
+            try:
+                await self._device_connect(location)
+            except UpnpError as err:
+                _LOGGER.warning(
+                    "Failed connecting to recently alive device at %s: %r",
+                    location,
+                    err,
+                )
+
+        # Device could have been de/re-connected, state probably changed
+        self.schedule_update_ha_state()
```

Comment:
We're not allowed to use the sync api inside the event loop. Please use `self.async_write_ha_state` instead.